compendium/DeclarativeServices/src/manager/ConfigurationManager.cpp

```diff
@@ -0,0 +1,172 @@
+/*=============================================================================
+
+  Library: CppMicroServices
+
+  Copyright (c) The CppMicroServices developers. See the COPYRIGHT
+  file at the top-level directory of this distribution and at
+  https://github.com/CppMicroServices/CppMicroServices/COPYRIGHT .
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  =============================================================================*/
+
+#include "ConfigurationManager.hpp"
+
+namespace cppmicroservices {
+namespace scrimpl {
+ConfigurationManager::ConfigurationManager(
+  const std::shared_ptr<const metadata::ComponentMetadata> metadata,
+  const cppmicroservices::BundleContext& bc,
+  std::shared_ptr<cppmicroservices::logservice::LogService> logger)
+  : metadata(metadata)
+  , logger(std::move(logger))
+  , bundleContext(bc)
+  , mergedProperties (metadata->properties)
+{
+  if (!this->metadata || !this->bundleContext || !this->logger) {
+    throw std::invalid_argument(
+      "ConfigurationManagerImpl - Invalid arguments passed to constructor");
+  }
+}
+
+cppmicroservices::AnyMap ConfigurationManager::GetProperties() const noexcept
+{
+  std::lock_guard<std::mutex> lock(propertiesMutex);
+
+  return mergedProperties;
+}
+
+void ConfigurationManager::Initialize() {
+  if ((metadata->configurationPids.empty()) ||
+      (metadata->configurationPolicy == metadata::ComponentMetadata::CONFIG_POLICY_IGNORE))
```

Comment:
what happens if the pids are empty and the config policy is set to "require"? Is this possible at runtime? or does the tooling error out at build time (i.e. via the DS code gen tool)?

Reply:
This combination (configuration-policy = "require", configuration-pid = [""],) will thrown an exception that will be logged by MetadataParserImplV1::ParseAndGetComponentsMetadata The component won't be loaded.

If just configuration-policy is present and configuration-pid is missing two warning messages are generated. One by the source code generator during the build  and one by MetadataParserImplV1::CreateComponentMetadata during runtime (which goes in the log). The component will be created with configurationPolicy = "ignore" in this case.  

Let me know if you want me to change the empty configuration-pid behavior to be consistent with the missing configuration-pid behavior.